pycrostates/cluster/aahc.py

```diff
@@ -0,0 +1,408 @@
+            self._labels_,
+            ignore_polarity=self._ignore_polarity,
+            normalize_input=self._normalize_input,
+            tol=self._tol,
+            GEV_=self._GEV_,
+        )
+
+    # --------------------------------------------------------------------
+    @staticmethod
+    def _aahc(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+        tol: Union[int, float],
+    ) -> Tuple[float, NDArray[float], NDArray[int]]:
+        """Run the AAHC algorithm."""
+        gfp_sum_sq = np.sum(data**2)
+        maps, segmentation = AAHCluster._compute_maps(
+            data, n_clusters, ignore_polarity, normalize_input, tol
+        )
+        map_corr = _corr_vectors(data, maps[segmentation].T)
+        gev = np.sum((data * map_corr) ** 2) / gfp_sum_sq
+        return gev, maps, segmentation
+
+    @staticmethod
+    def _compute_maps(
+        data: NDArray[float],
+        n_clusters: int,
+        ignore_polarity: bool,
+        normalize_input: bool,
+        tol: Union[int, float],
+    ) -> Tuple[NDArray[float], NDArray[int]]:
+        """Compute microstates maps."""
+        n_chan, n_frame = data.shape
+
+        cluster = data.copy()
+        cluster /= np.linalg.norm(cluster, axis=0, keepdims=True)
+
+        if normalize_input:
+            data = cluster.copy()
+
+        GEV = np.sum(data * cluster, axis=0)
+
+        assignment = np.arange(n_frame)
+
+        while cluster.shape[1] > n_clusters:
+
+            to_remove = np.argmin(GEV)
+            orphans = assignment == to_remove
+
+            old_cluster = cluster[:, to_remove].copy()
+
+            cluster = np.delete(cluster, to_remove, axis=1)
+            GEV = np.delete(GEV, to_remove, axis=0)
+            assignment[assignment > to_remove] = (
+                assignment[assignment > to_remove] - 1
+            )
+
+            fit = data[:, orphans].T @ cluster
+            if ignore_polarity:
+                fit = np.abs(fit)
+            new_assignment = np.argmax(fit, axis=1)
+            assignment[orphans] = new_assignment
+
+            cluster_to_update = np.unique(new_assignment)
+            for c in cluster_to_update:
+                members = assignment == c
+                if ignore_polarity:
+
+                    old_weight = len(new_assignment == c) / members.sum()
+
+                    sgn = np.sign(old_cluster @ cluster[:, c])
+
+                    v0 = old_weight * sgn * old_cluster + \
+                        (1. - old_weight) * cluster[:, c]
+
+                    v, _ = AAHCluster._first_principal_component(
+                        data[:, members], tol, v0
+                    )
+                    cluster[:, c] = v
+                else:
+                    cluster[:, c] = np.mean(data[:, members], axis=0)
+                    cluster[:, c] /= np.linalg.norm(
+                        cluster[:, c], axis=0, keepdims=True
+                    )
```

Comment:
I was looking at how `tol` was used, and how it defines convergence. Note: I am not familiar with AAHC.

It looks like:
- `tol` is not used if `ignore_polarity=False`. That's weird, right?
- There is no information/fix possible in case the convergence was not reached. I think a warning should be emitted if convergence was not reached and `AAHCluster._first_principal_component` has a `max_iter` argument which should be forwarded to the public API, to enable the user to increase it if need be.

Reply:
Thank you for the comment. Yes, `tol` was only used in the case `ignore_polarity=True`. In this case, AAHC requires to compute the first principal component (PC) of the cluster members. There are two options. First, power iterations (as implemented), which is fast but requires some heuristic to determine convergence. Second, conventional PCA which is slow because it requires the covariance matrix and then computes all eigenvectors/values.

In commit  https://github.com/vferat/pycrostates/pull/92/commits/a3f7afc062b2f43cf61a9bc2b43f7af29a470524, I removed the `tol` parameter from the public API. In case the fast power iteration method does not converge within a reasonable iteration (e.g., number of features = number of channels), the fallback method is used (covariance matrix + largest eigenvalue/vector).

Reply:
Thank you, looks great, I love this fallback ;)